### PR TITLE
fix: support lumigo edge latency [RD-3151]

### PR DESCRIPTION
**Pull request by doriaviram**

### CODE REVIEW

File: src/reporter.js [hunk 1]
```diff
@@ -25,33 +25,16 @@ export const isSpansContainsErrors = spans => {
 };
 
 export const sendSpans = async spans => {
-  const { token } = TracerGlobals.getTracerInputs();
```
Comment (saartochner) [inline src/reporter.js:28]: beautiful!

File: src/httpAgent.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { TracerGlobals } from './globals';
+import { getEdgeUrl, getJSONBase64Size, getTracerInfo } from './utils';
+import axios from 'axios';
+import http from 'http';
+import https from 'https';
+
+export const HttpAgent = (() => {
+  let sessionInstance = undefined;
+
+  const validateStatus = () => true;
```
[the remark this answers is not in the document]
Comment (doriaviram) [reply src/httpAgent.js:11]: Its mean, if the status code was 500+ any way consider this as a success request
the original implement is
  ```validateStatus: function (status) {
    return status >= 200 && status < 300; // default
  },```

File: src/tracer.test.js [hunk 3]
```diff
@@ -573,7 +574,7 @@ describe('tracer', () => {
     );
 
     expect(result[LUMIGO_EVENT_KEY][STEP_FUNCTION_UID_KEY]).not.toEqual('old');
-    const requests = HttpsRequestsForTesting.getRequests();
+    const requests = AxiosMocker.getAxiosMocker().history.post;
```
[the remark this answers is not in the document]
Comment (doriaviram) [reply src/tracer.test.js:577]: Fixing

File: src/reporter.test.js [hunk 4]
```diff
@@ -28,10 +14,8 @@ describe('reporter', () => {
 
     const result = await reporter.sendSingleSpan(span);
 
-    const requests = HttpsRequestsForTesting.getRequests();
-    const expectedRequest = buildExpectedRequest(token, [span]);
-
-    expect(requests).toEqual([expectedRequest]);
+    const spans = AxiosMocker.getSentSpans();
+    expect(spans).toEqual([[span]]);
```
[the remark this answers is not in the document]
Comment (doriaviram) [reply src/reporter.test.js:18]: List of requests, every request has list of spans

File: src/httpAgent.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+    return sessionInstance;
+  };
+
+  const postSpans = async requestBody => {
+    const session = getSessionInstance();
+    const bodySize = getJSONBase64Size(requestBody);
+    logger.debug('Starting request to edge', { bodySize });
+    await session.post('', requestBody).catch(e => {
+      logger.debug('Edge error (Tracer skipping)', e.message);
```
Comment (nirLumigo) [inline src/httpAgent.js:53]: Maybe it should be higher than `debug`?

File: src/httpAgent.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import * as logger from './logger';
+import { TracerGlobals } from './globals';
+import { getEdgeUrl, getJSONBase64Size, getTracerInfo } from './utils';
+import axios from 'axios';
+import http from 'http';
+import https from 'https';
+
+export const HttpAgent = (() => {
```
Comment (nirLumigo) [inline src/httpAgent.js:8]: As I understand, this file isn't relevant to HTTP hooks, it's only for reporting the spans. Maybe rename it so it will be easier to understand.

File: src/httpAgent.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      Authorization: token,
+      'User-Agent': `${name}$${version}`,
+      'Content-Type': 'application/json',
+    };
+  };
+
+  const createSessionInstance = (url, headers) => {
+    return axios.create({
+      baseURL: url,
+      timeout: 250,
```
Comment (nirLumigo) [inline src/httpAgent.js:26]: Why 250?

File: src/httpAgent.test.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    AxiosMocker.getAxiosMocker()
+      .onPost()
+      .networkError();
+
+    await HttpAgent.postSpans(reqBody);
+
+    //no Error throwed
+  });
+
+  test('postSpans - reject network error', async () => {
```
Comment (nirLumigo) [inline src/httpAgent.test.js:53]: Isn't this test the same as the test above?

File: testUtils/axiosMocker.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+  const createAxiosMocker = axiosLib => {
+    axiosMocker = new MockAdapter(axiosLib);
+  };
+
+  const getAxiosMocker = () => axiosMocker;
+
+  const getRequests = () => axiosMocker.history.post;
+
+  const getSentSpans = () => axiosMocker.history.post.map(r => JSON.parse(r.data));
```
Comment (nirLumigo) [inline testUtils/axiosMocker.js:14]:
```suggestion
  const getSentSpans = () => getRequests().map(r => JSON.parse(r.data));
```

File: testUtils/httpsMocker.js [hunk 10]
```diff
@@ -1,5 +1,7 @@
 const EventEmitter = require('events');
 
+class Agent {}
```
Comment (nirLumigo) [inline testUtils/httpsMocker.js:3]: Why is this needed?

File: src/httpAgent.js (hunk 5 above)
Comment (doriaviram) [reply src/httpAgent.js:53]: No, it's okay, we allow the edge to get a timeout without notifying us any problem

File: src/httpAgent.test.js (hunk 8 above)
Comment (doriaviram) [reply src/httpAgent.test.js:53]: No, the test above testing the response code, this a client error

File: testUtils/axiosMocker.js (hunk 9 above)
Comment (doriaviram) [reply testUtils/axiosMocker.js:14]: 👍

File: testUtils/httpsMocker.js (hunk 10 above)
Comment (doriaviram) [reply testUtils/httpsMocker.js:3]: The Axios mocker use that

File: src/httpAgent.js (hunk 7 above)
Comment (doriaviram) [reply src/httpAgent.js:26]: ?

File: src/httpAgent.test.js (hunk 8 above)
Comment (nirLumigo) [reply src/httpAgent.test.js]: image
Both of them check that there isn't an exception and they have the same name.
